Modules/constants.c

```diff
@@ -6,373 +6,180 @@
-#endif
-#ifdef LDAP_OPT_X_TLS_PROTOCOL_MIN
-  add_int(d,OPT_X_TLS_PROTOCOL_MIN);
-#endif
-#ifdef LDAP_OPT_X_TLS_PACKAGE
-  add_int(d,OPT_X_TLS_PACKAGE);
-#endif
-#endif
-  add_int(d,OPT_X_SASL_MECH);
-  add_int(d,OPT_X_SASL_REALM);
-  add_int(d,OPT_X_SASL_AUTHCID);
-  add_int(d,OPT_X_SASL_AUTHZID);
-  add_int(d,OPT_X_SASL_SSF);
-  add_int(d,OPT_X_SASL_SSF_EXTERNAL);
-  add_int(d,OPT_X_SASL_SECPROPS);
-  add_int(d,OPT_X_SASL_SSF_MIN);
-  add_int(d,OPT_X_SASL_SSF_MAX);
-#ifdef LDAP_OPT_X_SASL_NOCANON
-  add_int(d,OPT_X_SASL_NOCANON);
-#endif
-#ifdef LDAP_OPT_X_SASL_USERNAME
-  add_int(d,OPT_X_SASL_USERNAME);
-#endif
-#ifdef LDAP_OPT_CONNECT_ASYNC
-  add_int(d,OPT_CONNECT_ASYNC);
-#endif
-#ifdef LDAP_OPT_X_KEEPALIVE_IDLE
-  add_int(d,OPT_X_KEEPALIVE_IDLE);
-#endif
-#ifdef LDAP_OPT_X_KEEPALIVE_PROBES
-  add_int(d,OPT_X_KEEPALIVE_PROBES);
-#endif
-#ifdef LDAP_OPT_X_KEEPALIVE_INTERVAL
-  add_int(d,OPT_X_KEEPALIVE_INTERVAL);
-#endif
+#define LDAP_ERROR_OFFSET       -LDAP_ERROR_MIN
 
-  add_int(d,DN_FORMAT_LDAP);
-  add_int(d,DN_FORMAT_LDAPV3);
-  add_int(d,DN_FORMAT_LDAPV2);
-  add_int(d,DN_FORMAT_DCE);
-  add_int(d,DN_FORMAT_UFN);
-  add_int(d,DN_FORMAT_AD_CANONICAL);
-  /* add_int(d,DN_FORMAT_LBER); */    /* "for testing only" */
-  add_int(d,DN_FORMAT_MASK);
-  add_int(d,DN_PRETTY);
-  add_int(d,DN_SKIP);
-  add_int(d,DN_P_NOLEADTRAILSPACES);
-  add_int(d,DN_P_NOSPACEAFTERRDN);
-  add_int(d,DN_PEDANTIC);
-
-  add_int(d,AVA_NULL);
-  add_int(d,AVA_STRING);
-  add_int(d,AVA_BINARY);
-  add_int(d,AVA_NONPRINTABLE);
-  
-  /*add_int(d,OPT_ON);*/
-  obj = PyInt_FromLong(1);
-  PyDict_SetItemString( d, "OPT_ON", obj );
-  Py_DECREF(obj);
-  /*add_int(d,OPT_OFF);*/
-  obj = PyInt_FromLong(0);
-  PyDict_SetItemString( d, "OPT_OFF", obj );      
-  Py_DECREF(obj);
-  
-  add_int(d,OPT_SUCCESS);
-
-  /* XXX - these belong in errors.c */
-
-  add_int(d,URL_ERR_BADSCOPE);
-  add_int(d,URL_ERR_MEM);
-
-  /* add_int(d,LIBLDAP_R); */
-#ifdef HAVE_LIBLDAP_R
-  obj = PyInt_FromLong(1);
-#else
-  obj = PyInt_FromLong(0);
-#endif
-  PyDict_SetItemString( d, "LIBLDAP_R", obj );
-  Py_DECREF(obj);
+static PyObject* errobjects[ LDAP_ERROR_MAX-LDAP_ERROR_MIN+1 ];
 
-  /* add_int(d,SASL); */
-#ifdef HAVE_SASL
-  obj = PyInt_FromLong(1);
-#else
-  obj = PyInt_FromLong(0);
-#endif
-  PyDict_SetItemString( d, "SASL_AVAIL", obj );
-  Py_DECREF(obj);
 
-  /* add_int(d,TLS); */
-#ifdef HAVE_TLS
-  obj = PyInt_FromLong(1);
-#else
-  obj = PyInt_FromLong(0);
-#endif
-  PyDict_SetItemString( d, "TLS_AVAIL", obj );
-  Py_DECREF(obj);
+/* Convert a bare LDAP error number into an exception */
+PyObject*
+LDAPerr(int errnum)
+{
+  if (errnum >= LDAP_ERROR_MIN && errnum <= LDAP_ERROR_MAX)
+    PyErr_SetNone(errobjects[errnum+LDAP_ERROR_OFFSET]);
+  else
+    PyErr_SetObject(LDAPexception_class,
+    Py_BuildValue("{s:i}", "errnum", errnum));
+  return NULL;
+}
+
+/* Convert an LDAP error into an informative python exception */
+PyObject*
+LDAPerror( LDAP *l, char *msg )
```

Comment:
Is this a straight copy from the other file?

Reply:
Yes, except whitespace and a comment.

    git diff HEAD^{/Refactor}~:Modules/errors.c HEAD^{/Refactor}:Modules/constants.c

Reply:
+1

(cool git magic)